### ISSUE TITLE
Not deleting WAL after backup expired on PG10

### ISSUE THREAD

**Issue by Tostino** [pgbackrest/pgbackrest#459]
Spoke with Stephen Frost in the PG Slack channel about this.

I have a backup server which is setup to keep 3 backups.  It has been running since November on 1.25, and has successfully expired old backups, and incrementals I had.  The issue is, it is not expiring the WAL backups, and my server is now nearly full due to it.

I am running Postgres 10.1.

Stephen said that he wasn't seeing it on any of his 9.x clusters, but he actually noticed the same was happening on a 10.x cluster he looked at after we were talking about the issue.

Let me know if you need any more info.

**Comment by sfrost** [comment]
Hi there!

Can you try doing this:

pgbackrest --stanza=<stanza> --log-level-console=debug expire > logfile.txt

And then sending that to us?

Thanks!

**Comment by cmwshang** [comment]
And you are sure archive-retention option is set?

Cynthia Shang
www.crunchydata.com

**Comment by Tostino** [comment]
Yes I have the retention options set.  It expires the old backups just fine, the WALs just don't go away.

@sfrost requested I add my .info files, so here they are.

[backup.info.txt](https://github.com/pgbackrest/pgbackrest/files/1569373/backup.info.txt)
[archive.info.txt](https://github.com/pgbackrest/pgbackrest/files/1569374/archive.info.txt)

And here is the other info he requested:
```
root@Backup01:~# pgbackrest --stanza=main --log-level-console=debug expire > logfile.txt
root@Backup01:~# cat logfile.txt
2017-12-18 20:22:15.969 P00   INFO: expire command begin 1.25: --buffer-size=8388608 --db1-host=10.136.38.185 --log-level-console=debug --repo-path=/var/lib/pgbackrest --retention-full=3 --stanza=main
2017-12-18 20:22:15.979 P00  DEBUG:     Storage::Posix::Driver->new(): bFileSync = <true>, bPathSync = <true>
2017-12-18 20:22:15.979 P00  DEBUG:     Storage::Local->new(): bAllowTemp = <true>, hRule = [hash], lBufferMax = 8388608, oDriver = [object], strDefaultFileMode = <0640>, strDefaultPathMode = <0750>, strPathBase = /var/lib/pgbackrest, strTempExtension = pgbackrest.tmp
2017-12-18 20:22:15.979 P00  DEBUG:     Storage::Local->pathExists(): strPathExp =
2017-12-18 20:22:15.979 P00  DEBUG:     Storage::Local->pathExists=>: bExists = true
2017-12-18 20:22:15.980 P00  DEBUG:     Common::Lock::lockAcquire(): bFailOnNoLock = <true>, bRemote = <false>, strLockType = expire
2017-12-18 20:22:15.980 P00  DEBUG:     Storage::Local->pathCreate(): bCreateParent = true, bIgnoreExists = true, strMode = 770, strPathExp = /tmp/pgbackrest
2017-12-18 20:22:15.980 P00  DEBUG:     Common::Lock::lockAcquire=>: bResult = true
2017-12-18 20:22:15.980 P00  DEBUG:     Storage::Local->pathCreate(): bCreateParent = true, bIgnoreExists = true, strMode = 0770, strPathExp = /var/log/pgbackrest
2017-12-18 20:22:16.040 P00  DEBUG:     Backup::Info->new(): bIgnoreMissing = <false>, bLoad = <true>, bRequired = <true>, bValidate = <true>, oStorage = <[object]>, strBackupClusterPath = /var/lib/pgbackrest/backup/main
2017-12-18 20:22:16.040 P00  DEBUG:     Storage::Local->openRead(): bIgnoreMissing = true, rhyFilter = [undef], xFileExp = /var/lib/pgbackrest/backup/main/backup.info
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->reconstruct(): bRequired = <true>, bSave = <true>, iCatalogVersion = [undef], iControlVersion = [undef], strDbVersion = [undef], ullDbSysId = [undef]
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Common::backupRegExpGet(): bAnchor = <true>, bDifferential = true, bFull = true, bIncremental = true
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Common::backupRegExpGet=>: strRegExp = ^[0-9]{8}\-[0-9]{6}F(\_[0-9]{8}\-[0-9]{6}(D|I)){0,1}$
2017-12-18 20:22:16.104 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = <false>, strExpression = ^[0-9]{8}\-[0-9]{6}F(\_[0-9]{8}\-[0-9]{6}(D|I)){0,1}$, strPathExp = /var/lib/pgbackrest/backup/main, strSortOrder = <forward>
2017-12-18 20:22:16.104 P00  DEBUG:     Storage::Local->list=>: stryFileList = (20171201-161850F, 20171202-040001F, 20171216-040002F)
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->current(): strBackup = 20171201-161850F
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->current=>: bTest = true
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->current(): strBackup = 20171202-040001F
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->current=>: bTest = true
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->current(): strBackup = 20171216-040002F
2017-12-18 20:22:16.104 P00  DEBUG:     Backup::Info->current=>: bTest = true
2017-12-18 20:22:16.104 P00  DEBUG:     Storage::Local->pathExists(): strPathExp = /var/lib/pgbackrest/backup/main/20171201-161850F
2017-12-18 20:22:16.104 P00  DEBUG:     Storage::Local->pathExists=>: bExists = true
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->exists(): strFileExp = /var/lib/pgbackrest/backup/main/20171201-161850F/backup.manifest
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->exists=>: bExists = true
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->pathExists(): strPathExp = /var/lib/pgbackrest/backup/main/20171202-040001F
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->pathExists=>: bExists = true
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->exists(): strFileExp = /var/lib/pgbackrest/backup/main/20171202-040001F/backup.manifest
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->exists=>: bExists = true
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->pathExists(): strPathExp = /var/lib/pgbackrest/backup/main/20171216-040002F
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->pathExists=>: bExists = true
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->exists(): strFileExp = /var/lib/pgbackrest/backup/main/20171216-040002F/backup.manifest
2017-12-18 20:22:16.105 P00  DEBUG:     Storage::Local->exists=>: bExists = true
2017-12-18 20:22:16.105 P00  DEBUG:     Backup::Common::backupRegExpGet(): bAnchor = <true>, bDifferential = <false>, bFull = true, bIncremental = <false>
2017-12-18 20:22:16.105 P00  DEBUG:     Backup::Common::backupRegExpGet=>: strRegExp = ^[0-9]{8}\-[0-9]{6}F$
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->list(): strFilter = ^[0-9]{8}\-[0-9]{6}F$, strOrder = <forward>
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->list=>: stryBackup = (20171201-161850F, 20171202-040001F, 20171216-040002F)
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Common::backupRegExpGet(): bAnchor = <true>, bDifferential = true, bFull = true, bIncremental = true
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Common::backupRegExpGet=>: strRegExp = ^[0-9]{8}\-[0-9]{6}F(\_[0-9]{8}\-[0-9]{6}(D|I)){0,1}$
2017-12-18 20:22:16.106 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = <false>, strExpression = ^[0-9]{8}\-[0-9]{6}F(\_[0-9]{8}\-[0-9]{6}(D|I)){0,1}$, strPathExp = <REPO:BACKUP>, strSortOrder = reverse
2017-12-18 20:22:16.106 P00  DEBUG:     Storage::Local->list=>: stryFileList = (20171216-040002F, 20171202-040001F, 20171201-161850F)
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->current(): strBackup = 20171216-040002F
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->current=>: bTest = true
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->current(): strBackup = 20171202-040001F
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->current=>: bTest = true
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->current(): strBackup = 20171201-161850F
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->current=>: bTest = true
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Common::backupRegExpGet(): bAnchor = <true>, bDifferential = <false>, bFull = true, bIncremental = <false>
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Common::backupRegExpGet=>: strRegExp = ^[0-9]{8}\-[0-9]{6}F$
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->list(): strFilter = ^[0-9]{8}\-[0-9]{6}F$, strOrder = reverse
2017-12-18 20:22:16.106 P00  DEBUG:     Backup::Info->list=>: stryBackup = (20171216-040002F, 20171202-040001F, 20171201-161850F)
2017-12-18 20:22:16.106 P00  DEBUG:     Archive::Info->new(): bIgnoreMissing = <false>, bLoad = <true>, bRequired = true, strArchiveClusterPath = /var/lib/pgbackrest/archive/main
2017-12-18 20:22:16.107 P00  DEBUG:     Storage::Local->openRead(): bIgnoreMissing = true, rhyFilter = [undef], xFileExp = /var/lib/pgbackrest/archive/main/archive.info
2017-12-18 20:22:16.108 P00  DEBUG:     Storage::Local->list(): bIgnoreMissing = true, strExpression = ^[0-9]+\.[0-9]+-[0-9]+$, strPathExp = <REPO:ARCHIVE>, strSortOrder = <forward>
2017-12-18 20:22:16.108 P00  DEBUG:     Storage::Local->list=>: stryFileList = ()
2017-12-18 20:22:16.108 P00  DEBUG:     Common::Lock::lockRelease(): bFailOnNoLock = <true>
2017-12-18 20:22:16.108 P00  DEBUG:     Common::Exit::exitSafe(): iExitCode = 0, oException = [undef], strSignal = [undef]
2017-12-18 20:22:16.109 P00  DEBUG:     Protocol::Helper::protocolDestroy(): bComplete = true, iRemoteIdx = [undef], strRemoteType = [undef]
2017-12-18 20:22:16.109 P00  DEBUG:     Protocol::Helper::protocolDestroy=>: iExitStatus = 0
2017-12-18 20:22:16.109 P00  DEBUG:     Common::Lock::lockRelease(): bFailOnNoLock = false
2017-12-18 20:22:16.109 P00   INFO: expire command end: completed successfully
2017-12-18 20:22:16.109 P00  DEBUG:     Common::Exit::exitSafe=>: iExitCode = 0
```

**Comment by dwsteele** [comment]
OK - I see what's going on here.  The archive dir is being searched with ^[0-9]+\.[0-9]+-[0-9]+$ which does not work with Postgres 10.  This expression is only used for expiration so looks like everything else is working.  We'll get this fix out as soon as possible.

Meanwhile, you can manually delete any major WAL directories in /var/lib/pgbackrest/archive/main/10-1 that are *before* your oldest full backup to free up space.  If you send info output I'll identify the directory range for you.

**Comment by Tostino** [comment]
It's no problem, I can identify the old ones.  Thanks for jumping on this so quickly!

**Comment by dwsteele** [comment]
This has been fixed at 67c8f0dc5b8264f3 and we'll be releasing 1.X in the next day or so.

Thanks for reporting this and providing the logs!
